File: tests/cross_driver/node/bulk_smoke.js
// Cross-driver bulk-write smoke — Node (mongo-node-driver).
//
// Same workload as tests/cross_driver/go/bulk/main.go: one mixed
// bulkWrite that insert/update/upsert/replace/delete-walks across
// six docs, then assert the result counts and final collection
// state. node-driver's bulk command builder is a separate
// implementation from pymongo's, so wire-shape divergences trip
// here.

const { MongoClient } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

async function main() {
  const uri = process.env.MONGODB_URI;
  if (!uri) fail('MONGODB_URI not set');

  const client = new MongoClient(uri, { serverSelectionTimeoutMS: 5000 });
  await client.connect();
  try {
    const coll = client.db('bulk_xd').collection('c');
    await coll.drop().catch(() => {});

    await coll.insertMany([
      { _id: 1, kind: 'old' },
      { _id: 2, kind: 'old' },
    ]);

    const res = await coll.bulkWrite([
      { insertOne: { document: { _id: 3, kind: 'fresh' } } },
      { updateOne: { filter: { _id: 1 }, update: { $set: { kind: 'new' } } } },
      { updateMany: { filter: { kind: 'old' }, update: { $set: { kind: 'new' } } } },
      { replaceOne: { filter: { _id: 3 }, replacement: { _id: 3, kind: 'replaced' } } },
      {
        updateOne: {
          filter: { _id: 99 },
          update: { $set: { kind: 'upserted' } },
          upsert: true,
        },
      },
      { deleteOne: { filter: { _id: 2 } } },
    ]);

    if (res.insertedCount !== 1) fail(`insertedCount: got ${res.insertedCount}, want 1`);
    if (res.matchedCount !== 3) fail(`matchedCount: got ${res.matchedCount}, want 3`);
    if (res.modifiedCount !== 3) fail(`modifiedCount: got ${res.modifiedCount}, want 3`);
    if (res.upsertedCount !== 1) fail(`upsertedCount: got ${res.upsertedCount}, want 1`);
    if (res.deletedCount !== 1) fail(`deletedCount: got ${res.deletedCount}, want 1`);

    const got = await coll.find({}).sort({ _id: 1 }).toArray();
    if (got.length !== 3) fail(`final docs: got ${got.length}, want 3`);
    if (got[0]._id !== 1 || got[0].kind !== 'new')
      fail(`doc[0]: got ${JSON.stringify(got[0])}`);
    if (got[1]._id !== 3 || got[1].kind !== 'replaced')
      fail(`doc[1]: got ${JSON.stringify(got[1])}`);
    if (got[2]._id !== 99 || got[2].kind !== 'upserted')
      fail(`doc[2]: got ${JSON.stringify(got[2])}`);

    console.log('OK');
  } finally {
    await client.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));

File: tests/cross_driver/ruby/bulk_smoke.rb
# Cross-driver bulk-write smoke — Ruby (mongo-ruby-driver).
#
# One mixed bulk_write that insert / update / replace / upsert /
# delete-walks across six docs, then asserts the result counts and
# final collection state. The Ruby driver's bulk command builder is
# a separate implementation from pymongo's.
require "mongo"

def fail_with(msg)
  warn(msg)
  exit(1)
end

uri = ENV["MONGODB_URI"] or fail_with("MONGODB_URI not set")
Mongo::Logger.logger.level = Logger::FATAL

client = Mongo::Client.new(uri, database: "bulk_xd_ruby", server_selection_timeout: 5)
begin
  coll = client["c"]
  coll.drop

  coll.insert_many([
    {"_id" => 1, "kind" => "old"},
    {"_id" => 2, "kind" => "old"},
  ])

  res = coll.bulk_write([
    {insert_one: {"_id" => 3, "kind" => "fresh"}},
    {update_one: {filter: {"_id" => 1}, update: {"$set" => {"kind" => "new"}}}},
    {update_many: {filter: {"kind" => "old"}, update: {"$set" => {"kind" => "new"}}}},
    {replace_one: {filter: {"_id" => 3}, replacement: {"_id" => 3, "kind" => "replaced"}}},
    {update_one: {filter: {"_id" => 99}, update: {"$set" => {"kind" => "upserted"}}, upsert: true}},
    {delete_one: {filter: {"_id" => 2}}},
  ])

  fail_with("inserted_count: got #{res.inserted_count}, want 1") unless res.inserted_count == 1
  fail_with("matched_count: got #{res.matched_count}, want 3") unless res.matched_count == 3
  fail_with("modified_count: got #{res.modified_count}, want 3") unless res.modified_count == 3
  fail_with("upserted_count: got #{res.upserted_count}, want 1") unless res.upserted_count == 1
  fail_with("deleted_count: got #{res.deleted_count}, want 1") unless res.deleted_count == 1

  got = coll.find({}).sort(_id: 1).to_a
  fail_with("final docs: got #{got.length}, want 3") unless got.length == 3
  fail_with("doc[0]: #{got[0]}") unless got[0]["_id"] == 1 && got[0]["kind"] == "new"
  fail_with("doc[1]: #{got[1]}") unless got[1]["_id"] == 3 && got[1]["kind"] == "replaced"
  fail_with("doc[2]: #{got[2]}") unless got[2]["_id"] == 99 && got[2]["kind"] == "upserted"

  puts "OK"
ensure
  client.close
end
